### PR TITLE
Add flag dualSourceBlendDynamic to cbState

### PR DESCRIPTION
**Pull request by tywuAMD**
This PR is a revised version of https://github.com/GPUOpen-Drivers/llpc/pull/2467

CTS cases *dEQP-VK.shader_object.pipeline_interaction.\** enables dual source blend **dynamically** but only set one output color in the fragment shader. Thus. the test would crash since it tried to access **m_blendSources[1]** in **dualSourceSwizzle()** when **m_blendSources[1]** is empty.

To handle that case, add a flag **dualSourceBlendDynamic**. When **dualSourceBlendDynamic** is true i.e. dual source blend is dynamically set, check whether the fragment shader has a output at Location 0, Index 1 in SPIRVReader.

Affected CTS list:
_dEQP-VK.shader_object.pipeline_interaction.\*_

### CODE REVIEW

**Comment by qiaojbao** [comment]
ok to test

**Review by nhaehnle — COMMENTED** [review]
Thanks for revising, the general approach looks good to me.

One comment inline. In addition, please increment the minor interface version and add a corresponding bullet to the version history comment near the top of the file.

File: include/vkgcDefs.h [hunk 1]
```diff
@@ -1198,12 +1198,13 @@ struct GraphicsPipelineBuildInfo {
...
-  PipelineOptions options;    ///< Per pipeline tuning/debugging options
-  bool unlinked;              ///< True to build an "unlinked" half-pipeline ELF
-  bool dynamicVertexStride;   ///< Dynamic Vertex input Stride is enabled.
-  bool enableUberFetchShader; ///< Use uber fetch shader
-  bool enableEarlyCompile;    ///< Whether enable early compile
+  NggState nggState;           ///< NGG state used for tuning and debugging
+  PipelineOptions options;     ///< Per pipeline tuning/debugging options
+  bool unlinked;               ///< True to build an "unlinked" half-pipeline ELF
+  bool dynamicVertexStride;    ///< Dynamic Vertex input Stride is enabled.
+  bool dynamicDualSourceBlend; ///< Dynamic dual source blend mode is set.
```
[the remark this answers is not in the document]
Comment (linqun) [reply include/vkgcDefs.h:1205]: Please also include it in pipeline dump, hash calculation and vfx library.

**Comment by amdvlk-admin** [comment]
4ae9f5a82e8f65fe31a19e6724196fc8da176d03 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/llpc/context/llpcGraphicsContext.cpp:299:55: error: ‘const struct Vkgc::GraphicsPipelineBuildInfo’ has no member named ‘cbstate’; did you mean ‘cbState’?
       cbState.dualSourceBlendEnable || (pipelineInfo->cbstate.dualSourceBlendDynamic && getUseDualSourceBlend());
                                                       ^~~~~~~
                                                       cbState
[2806/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/context/llpcContext.cpp.o
[2807/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/context/llpcRayTracingContext.cpp.o

**Comment by amdvlk-admin** [comment]
6f66f83361c5018dce0dccce5fdb228f75d5d03f Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/llpc/context/llpcGraphicsContext.cpp:299:55: error: ‘const struct Vkgc::GraphicsPipelineBuildInfo’ has no member named ‘cbstate’; did you mean ‘cbState’?
       cbState.dualSourceBlendEnable || (pipelineInfo->cbstate.dualSourceBlendDynamic && getUseDualSourceBlend());
                                                       ^~~~~~~
                                                       cbState
[2797/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/context/llpcRayTracingContext.cpp.o
[2798/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/lower/llpcSpirvLowerInstMetaRemove.cpp.o

File: include/vkgcDefs.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tywuAMD) [reply include/vkgcDefs.h]: No problem. Just uploaded a new patch according to your comments.

**Comment by amdvlk-admin** [comment]
64a8b4b4a44cb07b49f16a892e4238520dc028b6 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/llpc/context/llpcGraphicsContext.cpp:299:55: error: ‘const struct Vkgc::GraphicsPipelineBuildInfo’ has no member named ‘cbstate’; did you mean ‘cbState’?
       cbState.dualSourceBlendEnable || (pipelineInfo->cbstate.dualSourceBlendDynamic && getUseDualSourceBlend());
                                                       ^~~~~~~
                                                       cbState
[2799/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/context/llpcContext.cpp.o
[2800/3065] Building CXX object compiler/llpc/CMakeFiles/llpcinternal.dir/lower/llpcSpirvLowerInstMetaRemove.cpp.o

**Review by linqun — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 38ac8766934fba104576cf47562e33ffbc6807bd
<details><summary>CTS tests (Failed: 0/213076)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 40426/71028 (56.9%)</li>
<li>Failed: 0/71028 (0.0%)</li>
<li>Not Supported: 30602/71028 (43.1%)</li><li>Warnings: 0/71028 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 40464/71018 (57.0%)</li>
<li>Failed: 0/71018 (0.0%)</li>
<li>Not Supported: 30554/71018 (43.0%)</li><li>Warnings: 0/71018 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.6, Gfx10</summary><ul>

<li>Passed: 36471/71030 (51.3%)</li>
<li>Failed: 0/71030 (0.0%)</li>
<li>Not Supported: 34559/71030 (48.7%)</li><li>Warnings: 0/71030 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**